File: 2479-meeting-rooms-iii/meeting-rooms-iii.cpp
class Solution {
public:
    int mostBooked(int n, vector<vector<int>>& meetings) {
        // sort wrt start time 
        // priority queue - priortize meetings with earliest end time and with smaller room number
        sort(meetings.begin(), meetings.end());

        priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>>usedRooms;
        priority_queue<int, vector<int>, greater<int>>unusedRooms;
        vector<int>numMeetings(n, 0);

        for(int i = 0; i < n; i++) unusedRooms.push(i);


        for(auto meeting : meetings){
            long long start = meeting[0]; int end = meeting[1];
            while(!usedRooms.empty() && usedRooms.top().first <= start){
                int room = usedRooms.top().second;
                usedRooms.pop(); unusedRooms.push(room);
            }

            if(!unusedRooms.empty()){
                usedRooms.push({end, unusedRooms.top()});
                numMeetings[unusedRooms.top()]++;
                unusedRooms.pop();
            }
            else{
                auto [time, r] = usedRooms.top();
                usedRooms.pop();
                usedRooms.push({time + end - start, r});
                numMeetings[r]++;
            }
        }


        int result = INT_MIN; int roomNum = -1;

        for(int i = 0; i < n; i++){
            cout<<numMeetings[i]<<" ";
            if(numMeetings[i] > result) {
                result = numMeetings[i]; roomNum = i;
            }
        }

        return roomNum;

        
    }
};

File: 2479-meeting-rooms-iii/meeting-rooms-iii.java
class Solution {
    public int mostBooked(int n, int[][] meetings) {
        int[] count = new int[n];
        Arrays.sort(meetings, (a, b) -> Integer.compare(a[0], b[0]));
        PriorityQueue<Integer> free = new PriorityQueue<>(); // min heap by default
        PriorityQueue<long[]> used = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]) != 0 ? Long.compare(a[0], b[0]) : Long.compare(a[1], b[1])); // [endTime, roomNum];

        for(int i = 0; i < n; i++) free.add(i);

        for(int[] meeting : meetings){
            int start = meeting[0]; int end = meeting[1];
            while(!used.isEmpty() && used.peek()[0] <= start){
                int usedRoom = (int) used.peek()[1]; used.poll(); free.add(usedRoom);
            }

            if(!free.isEmpty()){
                int room = free.poll();
                used.add(new long[]{end, room}); count[room]++;
            } else {
                long[] nxt = used.poll();
                used.add(new long[]{(end + nxt[0] - start), nxt[1]});
                count[(int)nxt[1]]++;
            }
        }

        int max = 0; int res= 0;
        for(int i = 0; i < n; i++){
            if(count[i] > max){
                res = i; max = count[i];
            }
        }

        return res;
    }
}